tests/cpp/test_matmul_scheduler.cpp

```diff
@@ -2248,6 +2248,104 @@ TEST_F(MatmulSchedulerTest, StridedBatchEpilogueSingleBias) {
   }
 }
 
+// Test matmul with sizes that are not divisible by 8 and with misaligned inputs
+TEST_F(MatmulSchedulerTest, MisalignedVectorization) {
+  // TODO: parametrized test instead of nested loops (still use a loop over
+  // sizes and re-use FusionExecutorCache)
+  for (auto layout : kAllSupportedMmaLayout) {
+    for (bool downcast_output : {false, true}) {
+      // TODO: maybe use bias
+      auto fusion = std::make_unique<Fusion>();
+      FusionGuard fg(fusion.get());
+
+      auto tv0 = makeContigTensor(2, DataType::Half);
+      auto tv1 = makeContigTensor(2, DataType::Half);
+      fusion->addInput(tv0);
+      fusion->addInput(tv1);
+
+      tv0 = canonicalizeInputToBMNK(tv0, layout, MmaOperand::A);
+      tv1 = canonicalizeInputToBMNK(tv1, layout, MmaOperand::B);
+      auto tv2 = fusedMultiplySum(tv0, tv1, {-1});
+
+      if (downcast_output) {
+        tv2 = castOp(DataType::Half, tv2);
+      }
+
+      fusion->addOutput(tv2);
+
+      NVF_CHECK(
+          1 == ir_utils::getOpsOfType<MmaOp>(fusion.get()).size(),
+          "matmul fusion must have at least one MmaOp");
+      NVF_CHECK(
+          ir_utils::getOpsOfType<MmaOp>(fusion.get())
+              .front()
+              ->layout()
+              .has_value(),
+          "input layout has not be set for MmaOp");
+      NVF_CHECK(
+          MmaLayout::TN ==
+              ir_utils::getOpsOfType<MmaOp>(fusion.get())
+                  .front()
+                  ->layout()
+                  .value(),
+          "the MmaOp layout of Ampere MMA must be always TN");
+
+      const auto fusion_layout = mma_utils::getMmaLayout(fusion.get());
+      NVF_CHECK(
+          fusion_layout.isValid(),
+          "failed to get decide matmul layout through fusion definition");
+      NVF_CHECK(
+          fusion_layout.getData() == layout,
+          "mismatch between test layout (",
+          toString(layout),
+          ") and layout inferred from fusion definition (",
+          toString(fusion_layout.getData()),
+          ")");
+
+      FusionExecutorCache executor_cache(std::move(fusion));
+
+      for (const auto& [M, N, K, alignA, alignB, alignBias] :
+           std::vector<std::tuple<int, int, int, int, int, int>>{
+               {504, 136, 248, 16, 16, 16}, // all fully vectorizable in all
+                                            // layouts
+               {504, 136, 249, 16, 16, 16}, // odd K, operands not vectorizable
```

Comment:
This case has odd K, which is the inner dimension for A in the TT and TN cases, and is the inner dimension for B in the NN and TN cases. So I expect either A or B to have a maximum supported vectorization of 1 and the output to have max vec of 4 for HSS and 8 for HSH. I am seeing 8 8 4 instead, in the TT case, so it seems there is more work to do on the analysis side.

Reply:
NN should be transposing `B`, but it seems `vectorize_helper` is not aware of that. Also, it doesn't assume we'll be doing vectorized loads starting at each row, instead of merging all contiguous dims then looping over that 1D dim. I think it's best to directly evaluate the stride of the M, N, and K dimensions in each operand (as well as the tensor pointer alignment and any outer strides as well).

Reply:
I think this means that we can look at each operand, find the dimension with the smallest stride, assert that that stride is 1, then find the largest power of two that's at most 16 and which divides evenly into all other strides. If we know the tensor is contiguous then I guess we could just do this for the smallest stride which is not 1 since all others are a multiple of it.

Reply:
Fixed in latest push. I no longer use `vectorize_helper` at all.